compiler/rustc_const_eval/src/interpret/operand.rs

```diff
@@ -730,14 +730,35 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                 // Return the cast value, and the index.
                 (discr_val, index.0)
             }
-            TagEncoding::Niche { untagged_variant, ref niche_variants, niche_start } => {
+            TagEncoding::Niche { untagged_variant, ref niche_variants, niche_start, ref flag } => {
+                let is_magic_val = if let Some(flag) = flag {
+                    let flag_val = self.read_immediate(&self.operand_field(op, flag.field)?)?;
+                    let flag_val = flag_val.to_scalar();
+                    match flag_val.try_to_int() {
+                        Err(_) => {
+                            // So this is a pointer then, and casting to an int failed. Can only
+                            // happen during CTFE.
+                            // The flag is not set to magic_value
```

Comment:
This is suspicious. We should probably abort execution here, I am fairly sure just silently continuing with a fallback will produce incorrect results.

Reply:
I'm not sure this is accurate. If the address is relative (`Provenance::OFFSET_IS_ADDR` is `false`), we still should be able to read the discriminant without just aborting. Essentially the same thing happens below already, in computing the discriminant for a regular old niche.

Reply:
But here you just assume that the pointer (when converted to an integer) is not equal to the magic value. What is the basis for that assumption?

Reply:
> But here you just assume that the pointer (when converted to an integer) is not equal to the magic value. What is the basis for that assumption?

The only possible magic value for a pointer is 0, and if the pointer fails to convert to an integer, it's not null. I will add a similar check to what appears below, but my understanding is that the check below is like a sanity check or assertion, not checking for something that is intended to occur in practice.

Reply:
Its' not an `assert!` because it can definitely happen in practice. You just need to write the right kind of `unsafe const` code to trigger it.

Reply:
[Here's a way to trigger that error](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f25cdb7cd38d79be3d5370a55d8803c9).

Reply:
>  if the pointer fails to convert to an integer, it's not null. 

That assumption is wrong. `(&0 as *const i32).wrapping_offset(0x10000)` will fail to convert to an integer, but might be null (depending on where that `0` will be allocated at runtime).